File: art.py
"""Programmatically drawn, animated sprites.

Every make_* function returns a list of pygame.Surface frames (with alpha).
Animation is achieved by cycling through the frames. No external art needed;
swap these out for real image assets later if desired.
"""
import math
import pygame
from settings import *


def _surf(size):
    return pygame.Surface((size, size), pygame.SRCALPHA)


def _bob(frame_i, frames, amp):
    """Vertical bob offset for a given frame (smooth sine)."""
    return int(math.sin(frame_i / frames * math.tau) * amp)


# ---------------------------------------------------------------- PLAYERS ----
def make_duck(size=48, frames=4):
    """Debug Duck: a rubber duck with a magnifying glass."""
    out = []
    for i in range(frames):
        s = _surf(size)
        bob = _bob(i, frames, 2)
        cx = size // 2
        body_y = size // 2 + bob
        # body
        pygame.draw.ellipse(s, YELLOW, (cx - 16, body_y - 8, 32, 26))
        # head
        pygame.draw.circle(s, YELLOW, (cx + 8, body_y - 10), 11)
        # beak
        pygame.draw.polygon(s, ORANGE, [(cx + 18, body_y - 12), (cx + 28, body_y - 9), (cx + 18, body_y - 6)])
        # eye
        pygame.draw.circle(s, BLACK, (cx + 11, body_y - 13), 2)
        # magnifying glass (debug tool)
        gx, gy = cx - 16, body_y + 8
        pygame.draw.circle(s, CYAN, (gx, gy), 6, 2)
        pygame.draw.line(s, DIM, (gx - 4, gy + 4), (gx - 10, gy + 10), 2)
        out.append(s)
    return out


def make_penguin(size=48, frames=4):
    """Linux Penguin (Tux-inspired)."""
    out = []
    for i in range(frames):
        s = _surf(size)
        bob = _bob(i, frames, 2)
        cx = size // 2
        by = size // 2 + bob
        # body (black)
        pygame.draw.ellipse(s, (20, 20, 20), (cx - 13, by - 14, 26, 34))
        # belly (white)
        pygame.draw.ellipse(s, WHITE, (cx - 9, by - 6, 18, 24))
        # head
        pygame.draw.circle(s, (20, 20, 20), (cx, by - 16), 10)
        # eyes
        pygame.draw.circle(s, WHITE, (cx - 3, by - 17), 3)
        pygame.draw.circle(s, WHITE, (cx + 3, by - 17), 3)
        pygame.draw.circle(s, BLACK, (cx - 3, by - 17), 1)
        pygame.draw.circle(s, BLACK, (cx + 3, by - 17), 1)
        # beak + feet (orange)
        pygame.draw.polygon(s, ORANGE, [(cx - 3, by - 12), (cx + 3, by - 12), (cx, by - 8)])
        footwig = 2 if i % 2 == 0 else -2
        pygame.draw.ellipse(s, ORANGE, (cx - 10, by + 17, 8, 5))
        pygame.draw.ellipse(s, ORANGE, (cx + 2 + footwig, by + 17, 8, 5))
        out.append(s)
    return out


def make_robot(size=48, frames=4):
    """Compiler-bot: a friendly boxy robot."""
    out = []
    for i in range(frames):
        s = _surf(size)
        bob = _bob(i, frames, 2)
        cx = size // 2
        by = size // 2 + bob
        # body
        pygame.draw.rect(s, (100, 116, 139), (cx - 12, by - 6, 24, 24), border_radius=5)
        # head
        pygame.draw.rect(s, (148, 163, 184), (cx - 10, by - 22, 20, 16), border_radius=4)
        # antenna
        pygame.draw.line(s, DIM, (cx, by - 22), (cx, by - 28), 2)
        pygame.draw.circle(s, RED, (cx, by - 29), 2)
        # eyes (blink on one frame)
        eye = CYAN if i != 2 else DIM
        pygame.draw.circle(s, eye, (cx - 5, by - 14), 3)
        pygame.draw.circle(s, eye, (cx + 5, by - 14), 3)
        # chest light
        pygame.draw.circle(s, GREEN, (cx, by + 4), 3)
        out.append(s)
    return out


PLAYER_MAKERS = {
    "duck": make_duck,
    "penguin": make_penguin,
    "robot": make_robot,
}


# ---------------------------------------------------------------- ENEMIES ----
def make_bug(size=40, frames=4, color=RED):
    """Generic 'bug' enemy with wiggling legs."""
    out = []
    for i in range(frames):
        s = _surf(size)
        cx = cy = size // 2
        legwig = 2 if i % 2 == 0 else -2
        for dx in (-10, 0, 10):
            pygame.draw.line(s, (30, 30, 30), (cx + dx // 2, cy),
                             (cx + dx, cy + 12 + legwig), 2)
            pygame.draw.line(s, (30, 30, 30), (cx + dx // 2, cy),
                             (cx + dx, cy - 12 - legwig), 2)
        pygame.draw.ellipse(s, color, (cx - 12, cy - 9, 24, 18))
        pygame.draw.circle(s, (20, 20, 20), (cx + 6, cy - 6), 3)
        pygame.draw.circle(s, WHITE, (cx - 3, cy - 2), 2)
        pygame.draw.circle(s, WHITE, (cx + 3, cy - 2), 2)
        out.append(s)
    return out


def make_shooter(size=44, frames=4):
    """Ranged enemy: a floating 'null-pointer' orb that fires."""
    out = []
    for i in range(frames):
        s = _surf(size)
        cx = cy = size // 2
        pulse = 2 + (i % 2) * 2
        pygame.draw.circle(s, PURPLE, (cx, cy), 13)
        pygame.draw.circle(s, (30, 10, 40), (cx, cy), 13, 2)
        pygame.draw.circle(s, CYAN, (cx, cy), 5 + pulse, 2)
        pygame.draw.circle(s, WHITE, (cx, cy), 3)
        out.append(s)
    return out


def make_tank(size=52, frames=4):
    """Tanky slow enemy: a 'memory-leak' blob."""
    out = []
    for i in range(frames):
        s = _surf(size)
        cx = cy = size // 2
        wob = _bob(i, frames, 2)
        pygame.draw.circle(s, ORANGE, (cx, cy + wob), 18)
        pygame.draw.circle(s, (120, 50, 10), (cx, cy + wob), 18, 3)
        pygame.draw.circle(s, YELLOW, (cx - 6, cy - 4 + wob), 3)
        pygame.draw.circle(s, YELLOW, (cx + 6, cy - 4 + wob), 3)
        pygame.draw.arc(s, (120, 50, 10), (cx - 8, cy + wob, 16, 12), 3.4, 6.0, 2)
        out.append(s)
    return out


def make_boss(size=120, frames=4):
    """Final boss: the Kernel Panic."""
    out = []
    for i in range(frames):
        s = _surf(size)
        cx = cy = size // 2
        wob = _bob(i, frames, 3)
        pygame.draw.circle(s, (40, 10, 20), (cx, cy + wob), 46)
        pygame.draw.circle(s, RED, (cx, cy + wob), 46, 4)
        # angry eyes
        pygame.draw.polygon(s, YELLOW, [(cx - 24, cy - 12 + wob), (cx - 6, cy - 4 + wob), (cx - 24, cy + wob)])
        pygame.draw.polygon(s, YELLOW, [(cx + 24, cy - 12 + wob), (cx + 6, cy - 4 + wob), (cx + 24, cy + wob)])
        # mouth
        pygame.draw.arc(s, YELLOW, (cx - 20, cy + 4 + wob, 40, 26), 3.4, 6.0, 4)
        # spikes
        for a in range(0, 360, 45):
            r = math.radians(a + i * 6)
            x1 = cx + math.cos(r) * 46
            y1 = cy + wob + math.sin(r) * 46
            x2 = cx + math.cos(r) * 58
            y2 = cy + wob + math.sin(r) * 58
            pygame.draw.line(s, RED, (x1, y1), (x2, y2), 4)
        out.append(s)
    return out


# Registry of enemy art makers by type name.
ENEMY_MAKERS = {
    "bug": lambda: make_bug(color=RED),
    "greenbug": lambda: make_bug(color=GREEN),
    "shooter": make_shooter,
    "tank": make_tank,
    "boss": make_boss,
}

File: OSGame/src/art.js
// Programmatically generated animated sprites (no art assets required).
// Each maker draws N frames into textures named `${key}${i}` and registers a
// looping animation `${key}_anim`. Mirrors the pygame `art.py` shapes.
import { COL } from "./config.js";

function bob(i, frames, amp) {
  return Math.round(Math.sin((i / frames) * Math.PI * 2) * amp);
}

// Draw one frame with a Graphics object, then bake to a texture.
function bake(scene, key, size, drawFn) {
  const g = scene.add.graphics();
  drawFn(g, size);
  g.generateTexture(key, size, size);
  g.destroy();
}

function registerAnim(scene, key, frameCount, rate = 8) {
  const frames = [];
  for (let i = 0; i < frameCount; i++) frames.push({ key: `${key}${i}` });
  if (!scene.anims.exists(`${key}_anim`)) {
    scene.anims.create({ key: `${key}_anim`, frames, frameRate: rate, repeat: -1 });
  }
}

// ---------------------------------------------------------------- players ----
function makeDuck(scene, size = 48, frames = 4) {
  for (let i = 0; i < frames; i++) {
    bake(scene, `duck${i}`, size, (g) => {
      const b = bob(i, frames, 2), cx = size / 2, by = size / 2 + b;
      g.fillStyle(COL.yellow, 1);
      g.fillEllipse(cx, by + 5, 32, 26);            // body
      g.fillCircle(cx + 8, by - 10, 11);            // head
      g.fillStyle(COL.orange, 1);
      g.fillTriangle(cx + 18, by - 12, cx + 28, by - 9, cx + 18, by - 6); // beak
      g.fillStyle(COL.black, 1);
      g.fillCircle(cx + 11, by - 13, 2);            // eye
      g.lineStyle(2, COL.cyan, 1);
      g.strokeCircle(cx - 16, by + 8, 6);           // magnifier glass
      g.lineStyle(2, COL.dim, 1);
      g.lineBetween(cx - 20, by + 12, cx - 26, by + 18); // handle
    });
  }
  registerAnim(scene, "duck", frames);
}

function makePenguin(scene, size = 48, frames = 4) {
  for (let i = 0; i < frames; i++) {
    bake(scene, `penguin${i}`, size, (g) => {
      const b = bob(i, frames, 2), cx = size / 2, by = size / 2 + b;
      g.fillStyle(0x141414, 1);
      g.fillEllipse(cx, by + 3, 26, 34);            // body
      g.fillStyle(COL.white, 1);
      g.fillEllipse(cx, by + 6, 18, 24);            // belly
      g.fillStyle(0x141414, 1);
      g.fillCircle(cx, by - 16, 10);                // head
      g.fillStyle(COL.white, 1);
      g.fillCircle(cx - 3, by - 17, 3);
      g.fillCircle(cx + 3, by - 17, 3);
      g.fillStyle(COL.black, 1);
      g.fillCircle(cx - 3, by - 17, 1);
      g.fillCircle(cx + 3, by - 17, 1);
      g.fillStyle(COL.orange, 1);
      g.fillTriangle(cx - 3, by - 12, cx + 3, by - 12, cx, by - 8); // beak
      const fw = i % 2 === 0 ? 2 : -2;
      g.fillEllipse(cx - 6, by + 20, 8, 5);
      g.fillEllipse(cx + 6 + fw, by + 20, 8, 5);
    });
  }
  registerAnim(scene, "penguin", frames);
}

function makeRobot(scene, size = 48, frames = 4) {
  for (let i = 0; i < frames; i++) {
    bake(scene, `robot${i}`, size, (g) => {
      const b = bob(i, frames, 2), cx = size / 2, by = size / 2 + b;
      g.fillStyle(0x64748b, 1);
      g.fillRoundedRect(cx - 12, by - 6, 24, 24, 5);  // body
      g.fillStyle(0x94a3b8, 1);
      g.fillRoundedRect(cx - 10, by - 22, 20, 16, 4); // head
      g.lineStyle(2, COL.dim, 1);
      g.lineBetween(cx, by - 22, cx, by - 28);        // antenna
      g.fillStyle(COL.red, 1);
      g.fillCircle(cx, by - 29, 2);
      const eye = i !== 2 ? COL.cyan : COL.dim;
      g.fillStyle(eye, 1);
      g.fillCircle(cx - 5, by - 14, 3);
      g.fillCircle(cx + 5, by - 14, 3);
      g.fillStyle(COL.green, 1);
      g.fillCircle(cx, by + 4, 3);
    });
  }
  registerAnim(scene, "robot", frames);
}

// ---------------------------------------------------------------- enemies ----
function makeBug(scene, key, color, size = 40, frames = 4) {
  for (let i = 0; i < frames; i++) {
    bake(scene, `${key}${i}`, size, (g) => {
      const cx = size / 2, cy = size / 2;
      const lw = i % 2 === 0 ? 2 : -2;
      g.lineStyle(2, 0x1e1e1e, 1);
      for (const dx of [-10, 0, 10]) {
        g.lineBetween(cx + dx / 2, cy, cx + dx, cy + 12 + lw);
        g.lineBetween(cx + dx / 2, cy, cx + dx, cy - 12 - lw);
      }
      g.fillStyle(color, 1);
      g.fillEllipse(cx, cy, 24, 18);
      g.fillStyle(0x141414, 1);
      g.fillCircle(cx + 6, cy - 6, 3);
      g.fillStyle(COL.white, 1);
      g.fillCircle(cx - 3, cy - 2, 2);
      g.fillCircle(cx + 3, cy - 2, 2);
    });
  }
  registerAnim(scene, key, frames);
}

function makeShooter(scene, size = 44, frames = 4) {
  for (let i = 0; i < frames; i++) {
    bake(scene, `shooter${i}`, size, (g) => {
      const cx = size / 2, cy = size / 2, pulse = 2 + (i % 2) * 2;
      g.fillStyle(COL.purple, 1);
      g.fillCircle(cx, cy, 13);
      g.lineStyle(2, 0x1e0a28, 1);
      g.strokeCircle(cx, cy, 13);
      g.lineStyle(2, COL.cyan, 1);
      g.strokeCircle(cx, cy, 5 + pulse);
      g.fillStyle(COL.white, 1);
      g.fillCircle(cx, cy, 3);
    });
  }
  registerAnim(scene, "shooter", frames);
}

function makeTank(scene, size = 52, frames = 4) {
  for (let i = 0; i < frames; i++) {
    bake(scene, `tank${i}`, size, (g) => {
      const cx = size / 2, cy = size / 2, w = bob(i, frames, 2);
      g.fillStyle(COL.orange, 1);
      g.fillCircle(cx, cy + w, 18);
      g.lineStyle(3, 0x78320a, 1);
      g.strokeCircle(cx, cy + w, 18);
      g.fillStyle(COL.yellow, 1);
      g.fillCircle(cx - 6, cy - 4 + w, 3);
      g.fillCircle(cx + 6, cy - 4 + w, 3);
    });
  }
  registerAnim(scene, "tank", frames);
}

function makeBoss(scene, size = 128, frames = 4) {
  for (let i = 0; i < frames; i++) {
    bake(scene, `boss${i}`, size, (g) => {
      const cx = size / 2, cy = size / 2, w = bob(i, frames, 3);
      g.fillStyle(0x280a14, 1);
      g.fillCircle(cx, cy + w, 46);
      g.lineStyle(4, COL.red, 1);
      g.strokeCircle(cx, cy + w, 46);
      g.fillStyle(COL.yellow, 1);
      g.fillTriangle(cx - 24, cy - 12 + w, cx - 6, cy - 4 + w, cx - 24, cy + w);
      g.fillTriangle(cx + 24, cy - 12 + w, cx + 6, cy - 4 + w, cx + 24, cy + w);
      g.lineStyle(4, COL.red, 1);
      for (let a = 0; a < 360; a += 45) {
        const r = (a + i * 6) * Math.PI / 180;
        g.lineBetween(cx + Math.cos(r) * 46, cy + w + Math.sin(r) * 46,
                      cx + Math.cos(r) * 58, cy + w + Math.sin(r) * 58);
      }
    });
  }
  registerAnim(scene, "boss", frames, 6);
}

// Bullet textures (simple circles).
function makeBullet(scene, key, color, r = 6) {
  bake(scene, key, (r + 2) * 2, (g) => {
    g.fillStyle(color, 1);
    g.fillCircle(r + 2, r + 2, r);
    g.fillStyle(COL.white, 1);
    g.fillCircle(r + 2, r + 2, Math.max(2, r - 3));
  });
}

export function generateAllTextures(scene) {
  makeDuck(scene);
  makePenguin(scene);
  makeRobot(scene);
  makeBug(scene, "bug", COL.red);
  makeBug(scene, "greenbug", COL.green);
  makeShooter(scene);
  makeTank(scene);
  makeBoss(scene);
  makeBullet(scene, "pbullet", COL.cyan);
  makeBullet(scene, "ebullet", COL.purple);
  makeBullet(scene, "drone", COL.cyan, 8);
}
